File: HandSpecifications/FlushSpecification.php
<?php

class FlushSpecification extends HandSpecification {

    /**
     * @param Hand $Hand
     * @return boolean
     */
    public function isSatisfiedBy(Hand $Hand) {
        return $this->_canBeAFlush($Hand);
    }

    /**
     * @param Hand $Hand
     * @return bool
     */
    protected function _canBeAFlush(Hand $Hand) {
        return $this->newHand($Hand) instanceof Flush;
    }


    /**
     * @param Hand $Hand
     * @return Flush
     */
    public function newHand(Hand $Hand) {
        $SpadesCards = array();
        $HeartsCards = array();
        $ClubsCards = array();
        $DiamondsCards = array();

        foreach ($Hand->getCards() as $Card) {
            switch ($Card->getSuit()) {
                case Suit::SPADES:
                    $SpadesCards[] = $Card;
                    break;

                case Suit::HEARTS:
                    $HeartsCards[] = $Card;
                    break;

                case Suit::CLUBS:
                    $ClubsCards[] = $Card;
                    break;

                case Suit::DIAMONDS:
                    $DiamondsCards[] = $Card;
                    break;
            }

            foreach (array($SpadesCards, $HeartsCards, $ClubsCards, $DiamondsCards) as $Cards) {
                if (count($Cards) == 5) {
                    return new Flush($Cards);
                }
            }
        }

        return null;
    }
}

File: HandSpecifications/FlushSpecification.cs

class FlushSpecification : HandSpecification {

    public override bool isSatisfiedBy(Hand Hand) {
        return newHand(Hand) is Flush;
    }

    public override Hand newHand(Hand Hand) {
       var SpadesCards = new List<Card>();
       var HeartsCards = new List<Card>();
       var ClubsCards = new List<Card>();
       var DiamondsCards = new List<Card>();

        foreach (var Card in Hand.getCards()) {
            switch (Card.getSuit()) {
                case Suit::SPADES:
                    SpadesCards.Add(Card);
                    if (SpadesCards.Length() == 5) {
                        return new Flush(SpadesCards);
                    }
                    break;

                case Suit::HEARTS:
                    HeartsCards.Add(Card);
                    if (HeartsCards.Length() == 5) {
                        return new Flush(HeartsCards);
                    }
                    break;

                case Suit::CLUBS:
                    ClubsCards.Add(Card);
                    if (ClubsCards.Length() == 5) {
                        return new Flush(ClubsCards);
                    }
                    break;

                case Suit::DIAMONDS:
                    DiamondsCards.Add(Card);
                    if (DiamondsCards.Length() == 5) {
                        return new Flush(DiamondsCards);
                    }
                    break;
            }
        }

        return null;
    }
}